src/main/java/com/commercetools/payment/notification/CommercetoolsPaymentNotificationController.java

```diff
@@ -1,34 +1,79 @@
 
-import javax.servlet.http.HttpServletRequest;
-import java.util.Arrays;
-import java.util.stream.Collectors;
+import javax.annotation.Nonnull;
+import java.util.concurrent.CompletableFuture;
 
+import static com.commercetools.payment.constants.Psp.NOTIFICATION_PATH_URL;
 import static com.commercetools.payment.constants.Psp.PSP_NAME;
 import static java.lang.String.format;
+import static org.springframework.http.MediaType.APPLICATION_JSON_VALUE;
 
+/**
+ * Process the valid notification event from Paypal Plus. Validation is done in
+ * {@link NotificationValidationInterceptor}.
+ * 
+ * For devs: validation of Paypal notification cannot be done here because it needs HTTP request headers and body. However,
+ * {@link CommercetoolsPaymentNotificationController#handleNotification(String, PaypalPlusNotificationEvent)} also requires
+ * RequestBody to parse the Event object. In Spring, you cannot inject both RequestBody and HttpRequest in one method
+ * and then read the request body. It will throw an error.
+ */
 @RestController
-public class CommercetoolsPaymentNotificationController {
+public class CommercetoolsPaymentNotificationController extends BaseCommercetoolsPaymentsController {
 
-    @RequestMapping(
-            method = RequestMethod.POST,
-            value = "/{tenantName}/"+ PSP_NAME + "/notification")
-    public String handlePayments(@PathVariable String tenantName,
-                                 HttpServletRequest request) {
+    private final NotificationEventDispatcherProvider eventDispatcherProvider;
 
+    private final Logger logger;
 
-        // skeleton for tests: just "reflect" the tenant name and request arguments
-        String requestParameters = request.getParameterMap().entrySet().stream()
-                .map(stringEntry -> format("%s=%s", stringEntry.getKey(), Arrays.stream(stringEntry.getValue())
-                        .findFirst().orElse("")))
-                .collect(Collectors.joining("\n"));
+    @Autowired
+    public CommercetoolsPaymentNotificationController(@Nonnull StringTrimmerEditor stringTrimmerEditor,
+                                                      @Nonnull NotificationEventDispatcherProvider eventDispatcherProvider) {
+        super(stringTrimmerEditor);
+        this.eventDispatcherProvider = eventDispatcherProvider;
+        logger = LoggerFactory.getLogger(PaymentHandler.class);
+    }
 
-        return format("Requested tenant [%s] with data:%n%s%n", tenantName,
-                requestParameters);
+    @RequestMapping(
+            method = RequestMethod.POST,
+            consumes = APPLICATION_JSON_VALUE,
+            value = "/{tenantName}/" + NOTIFICATION_PATH_URL)
+    public ResponseEntity<HttpStatus> handleNotification(@PathVariable String tenantName,
+                                                         @RequestBody PaypalPlusNotificationEvent eventFromPaypal) {
+        return eventDispatcherProvider.getNotificationDispatcher(tenantName)
```

Comment:
pls, follow the convention in all other controllers:
  - carry out the business logic to a dedicated handler
  - return `PaymentHandleResponse` as in other handlers

Reply:
But the logic here is to generate the response, so it should be in the controller?

Reply:
yes, but we are trying to modularize the application, so we are not overloading _interface_ classes (like controllers, models) with heavy logic. 

We already started to do this in other controllers (see `CommercetoolsCreatePaymentsController`, `CommercetoolsExecutePaymentsController`). They look nice and easy to read. When you read the sources in those classes, you easily find the steps, which are done in them (like: verify request, handle request, convert result to response type). And any of these classes could be easily used as a wireframe to new controllers.

With current `CommercetoolsPaymentNotificationController` implementation i'm confused, which parts are just common REST controller specific, and which parts are notification specific. Because the modularization now lost.

So, pls, follow the convention we started from the beginning of the project. It makes our future life easier.

Reply:
https://github.com/commercetools/commercetools-paypal-plus-integration/pull/57/commits/fcabee91462a48d3cf06f020fae4404e360d941f

Reply:
https://github.com/commercetools/commercetools-paypal-plus-integration/pull/57/commits/65e4b0c1a246658912199d625b824630d5f4fa68